Svc/BufferManager/BufferManagerComponentImpl.cpp

```diff
@@ -0,0 +1,222 @@
+      FW_ASSERT(id < this->m_numStructs,id,this->m_numStructs);
+      FW_ASSERT(mgrId == this->m_mgrId,mgrId,id,this->m_mgrId);
+      FW_ASSERT(true == this->m_buffers[id].allocated,id,this->m_mgrId);
+      FW_ASSERT(reinterpret_cast<U8*>(fwBuffer.getData()) == this->m_buffers[id].memory,id,this->m_mgrId);
+      // user can make smaller for their own purposes, but it shouldn't be bigger
+      FW_ASSERT(fwBuffer.getSize() <= this->m_buffers[id].size,id,this->m_mgrId);
+      // clear the allocated flag
+      this->m_buffers[id].allocated = false;
+      this->m_currBuffs--;
+  }
+
+  Fw::Buffer BufferManagerComponentImpl ::
+    bufferGetCallee_handler(
+        const NATIVE_INT_TYPE portNum,
+        U32 size
+    )
+  {
+      // make sure component has been set up
+      FW_ASSERT(this->m_setup);
+      // find smallest buffer based on size.
+      for (NATIVE_UINT_TYPE buff = 0; buff < this->m_numStructs; buff++) {
+          if ((not this->m_buffers[buff].allocated) and (size < this->m_buffers[buff].size)) {
+              this->m_buffers[buff].allocated = true;
+              this->m_currBuffs++;
+              if (this->m_currBuffs > this->m_highWater) {
+                  this->m_highWater = this->m_currBuffs;
+              }
+              Fw::Buffer copy = this->m_buffers[buff].buff;
+              // change size to match request
+              copy.setSize(size);
+              return copy;
+          }
+      }
+
+      // if no buffers found, return empty buffer
+      this->log_WARNING_HI_NoBuffsAvailable(size);
+      this->m_noBuffs++;
+      return Fw::Buffer();
+
+  }
+
+  void BufferManagerComponentImpl::setup(
+    NATIVE_UINT_TYPE mgrId, //!< manager ID
+    NATIVE_UINT_TYPE memId, //!< Memory segment identifier
+    Fw::MemAllocator& allocator, //!< memory allocator
+    const BufferBins& bins //!< Set of user bins
+  ) {
+
+    this->m_mgrId = mgrId;
+    this->m_memId = memId;
+    this->m_allocator = &allocator;
+    // clear bins
+    memset(&this->m_bufferBins,0,sizeof(this->m_bufferBins));
+
+    this->m_bufferBins = bins;
+
+    // compute the amount of memory needed
+    NATIVE_UINT_TYPE memorySize = 0; // track needed memory
+    this->m_numStructs = 0; // size the number of tracking structs
+    // walk through bins and add up the sizes
+    for (NATIVE_UINT_TYPE bin = 0; bin < BUFFERMGR_MAX_NUM_BINS; bin++) {
+        if (this->m_bufferBins.bins[bin].numBuffers) {
+            memorySize += 
+                this->m_bufferBins.bins[bin].bufferSize * this->m_bufferBins.bins[bin].numBuffers // allocate each set of buffer memory
+                + sizeof(AllocatedBuffer) * this->m_bufferBins.bins[bin].numBuffers; // allocate the structs to track the buffers
+            this->m_numStructs += this->m_bufferBins.bins[bin].numBuffers;
+        }
+    }
+
+    NATIVE_UINT_TYPE allocatedSize = memorySize;
+    bool recoverable; //!< don't care if it is recoverable since they are a pool of user buffers
```

Comment:
This variable should be initialized to a specific value (true or false), even if the value doesn't matter. Pick one. Also, we should run static analysis and make sure there are no more uninitialized variables.

Reply:
This issue is resolved in pull request #338